romanisim/l3.py

```diff
@@ -17,64 +36,63 @@ def add_objects_to_l3(l3_mos, source_cat, rng=None, seed=None):
-
-    # Generate x,y positions for sources
-    coords = np.array([[o.sky_pos.ra.rad, o.sky_pos.dec.rad]
-                       for o in source_cat])
-    sourcecountsall = galsim.ImageF(l3_mos.data.shape[0], l3_mos.data.shape[1], wcs=twcs, xmin=0, ymin=0)
-    xpos, ypos = sourcecountsall.wcs._xy(coords[:, 0], coords[:, 1])
-    xpos_idx = [round(x) for x in xpos]
-    ypos_idx = [round(y) for y in ypos]
-
-    # Create overall scaling factor map
-    Ct_all = (l3_mos.data.value / l3_mos.var_poisson)
-
-    # Cycle over sources and add them to the mosaic
-    for idx, (x, y) in enumerate(zip(xpos_idx, ypos_idx)):
-        # Set scaling factor for injected sources
-        # Flux / sigma_p^2
-        Ct = (l3_mos.data[x][y].value / l3_mos.var_poisson[x][y].value)
-
-        # Create empty postage stamp galsim source image
-        sourcecounts = galsim.ImageF(l3_mos.data.shape[0], l3_mos.data.shape[1], wcs=twcs, xmin=0, ymin=0)
-
-        # Simulate source postage stamp
-        image.add_objects_to_image(sourcecounts, [source_cat[idx]], xpos=[xpos[idx]], ypos=[ypos[idx]],
-                                   psf=l3_psf, flux_to_counts_factor=Ct, bandpass=[filter_name],
-                                   filter_name=filter_name, rng=rng)
-
-        # Scale the source image back by its flux ratios
-        sourcecounts /= Ct
-
-        # Add sources to the original mosaic data array
-        l3_mos.data = (l3_mos.data.value + sourcecounts.array) * l3_mos.data.unit
-
-        # Note for the future - other noise sources (read and flat) need to be implemented
-
-    # Set new poisson variance
-    l3_mos.var_poisson = l3_mos.data.value / Ct_all
+    # Generate WCS (if needed)
+    if wcs is None:
+        wcs = romanisim.wcs.get_mosaic_wcs(l3_mos.meta, shape=l3_mos.data.shape)
+
+    # Create PSF (if needed)
+    if psf is None:
+        psf = romanisim.psf.make_psf(filter_name=filter_name, sca=CENTER_SCA, chromatic=False, webbpsf=True)
+
+    # Create Image canvas to add objects to
+    sourcecountsall = galsim.ImageF(l3_mos.data.value, wcs=wcs, xmin=0, ymin=0)
+
+    # Create position arrays (if needed)
+    if any(pos is None for pos in [xpos, ypos]):
+        # Create coordinates (if needed)
+        if coords is None:
+            coords = np.array([[o.sky_pos.ra.rad, o.sky_pos.dec.rad]
+                               for o in source_cat])
+            coords_unit = 'rad'
+        # Generate x,y positions for sources
+        xpos, ypos = sourcecountsall.wcs.radecToxy(coords[:, 0], coords[:, 1], coords_unit)
+
+    # Add sources to the original mosaic data array
+    romanisim.image.add_objects_to_image(sourcecountsall, source_cat, xpos=xpos, ypos=ypos,
+                                         psf=psf, flux_to_counts_factor=[xpt * unit_factor for xpt in exptimes],
+                                         exptimes=exptimes, bandpass=[filter_name], filter_name=filter_name,
+                                         wcs=wcs, rng=rng, seed=seed)
```

Comment:
Our naming isn't quite right, but the flux_to_counts_factor needs to take the source catalog fluxes (maggies in the simulator) into counts.  That factor will be the AB flux for the particular filter times the exposure time; that's close to what you have.  Then the "exptimes" factor is what we're using to go back from counts to whatever units the L3 file is in; MJy / sr.  That factor needs to be:
abflux * 3631 * 10^6 / (pix / steradian)
if I got it all right.  abflux puts one back in maggies / pixel units.  A maggie is the same as 3631 Jy, so then one needs to divide by 3631, convert from Jy to MJy, and and convert from pixels to steradians.  Let's just use a pixel in the center of the mosaic, though an argument could be made for computing the size of each pixel individually.

Reply:
Implemented.